File: application/src/main/java/com/catalog/application/category/retrieve/get/DefaultGetCategoryByIdUseCase.java
package com.catalog.application.category.retrieve.get;

import com.catalog.domain.category.Category;
import com.catalog.domain.category.CategoryGateway;
import com.catalog.domain.category.CategoryID;
import com.catalog.domain.exceptions.DomainException;
import com.catalog.domain.exceptions.NotFoundException;
import com.catalog.domain.validation.Error;

import java.util.Objects;
import java.util.function.Supplier;

public class DefaultGetCategoryByIdUseCase extends GetCategoryByIdUseCase {
    private final CategoryGateway categoryGateway;

    public DefaultGetCategoryByIdUseCase(final CategoryGateway categoryGateway) {
        this.categoryGateway = Objects.requireNonNull(categoryGateway);
    }

    @Override
    public CategoryOutput execute(final String anIn) {
        final var anId = CategoryID.from(anIn);
        return this.categoryGateway.findById(anId)
                .map(CategoryOutput::from)
                .orElseThrow(notFound(anId));
    }

    private Supplier<NotFoundException> notFound(final CategoryID anId) {
        return () -> NotFoundException.with(Category.class, anId);
    }
}

File: application/src/main/kotlin/category/retrieve/get/DefaultGetCategoryByIdUseCase.kt
package com.full.admin.catalog.application.category.retrieve.get

import category.CategoryGateway
import category.CategoryID
import exceptions.DomainException
import validation.Error

class DefaultGetCategoryByIdUseCase(private val categoryGateway: CategoryGateway) : GetCategoryByIdUseCase() {
    override fun execute(aCommand: String): CategoryOutput {
        val aCategoryID = CategoryID.from(aCommand)

        return this.categoryGateway.findById(aCategoryID)?.let(CategoryOutput::from) ?: throw DomainException.with(Error("Category with ID $aCommand was not found"))
    }
}